EIPS/eip-6963.md

```diff
@@ -0,0 +1,211 @@
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-05-01
+requires: 1193
+---
+
+## Abstract
+
+A Javascript communication protocol that enables multiple wallet providers to interface with an EVM-based application using a browser extension injected script through window events.
+
+## Motivation
+
+Currently, wallet providers that offer browser extensions must inject their Ethereum providers ([EIP-1193](./eip-1193.md)) into the same window object `window.ethereum` however this creates conflicts for users that may install more than one browser extension.
+
+Browser extensions are loaded in the web page and do not have a predictable pattern resulting in a race condition where the user is not in control to choose which wallet provider must be selected for exposing Ehtereum interface under the `window.ethereum` object.
+
+This results not only in a degraded user experience but also increases the barrier to entry for new browser extensions as users are forced to only install one browser extension at a time.
+
+Some browser extensions attempt to counteract this problem by delaying their injection to overwrite the same `window.ethereum` object which creates an unfair competition for wallet providers and lack of interoperability.
+
+In this proposal, we provide a solution that optimitizes for interoperability of multiple wallet providers and enables fairer competition by lowering the barrier to entry for new wallet providers to improve the user experience on Ethereum.
+
+This is achieved by introducing a set of window events to provide a two-way communication protocol between Ethereum libraries and injected scripts provided by browser extensions thus enabling users to select their wallet of choice.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Provider Info
+
+Each wallet provider will be exposed with the following interface `ProviderInfo` that will be used to display to the user:
+
+```typescript
+/**
+ * Represents the assets needed to display a wallet
+ */
+interface ProviderInfo {
+  walletId: string;
+  uuid: string;
+  name: string;
+  icon: string;
+}
+```
+
+The values in the `ProviderInfo` MUST be used as follows:
+
+- walletId - globally unique identifier of the wallet provider (eg. `io.dopewallet.extension` or `awesomewallet`)
+- uuid - locally unique of the wallet provider (UUIDv4 compliant)
+- name - human-readable name of the wallet provider (e.g. `DopeWalletExtension` or `AwesomeWallet`)
+- icon - uri encoded image (RFC-3986 complaint)
+
+### EVMProvider
+
+The wallet provider will also expose their own [EIP-1193](./eip-1193.md) provider interface in parallel with the provider info provided above in the following interface `EVMProvider`
+
+```typescript
+interface EVMProvider {
+  info: ProviderInfo;
+  ethereum: EIP1193Provider;
```

Comment:
Should this be attributed as `provider`? Kinda contradicts itself being `ethereum` when it represents an "evm provider", and feels like it's a provider within a provider lol.

Thought: align the type of `EVMProvider` closer to event terminology. If we are rolling with `MessageEvent`, this could be `EVMEventData`:

```ts
interface EVMEventData {
  info: ProviderInfo;
  provider: EIP1193Provider;
}
```

This represents itself much more clearer in implementation:

```ts
window.postMessage(
  new MessageEvent<EVMEventData>('evmProviderAnnounced', {
    data: {
      info: {
        walletId: 'org.example.wallet',
        uuid: '350670db-19fa-4704-a166-e52e178b59d2',
        name: 'Example Wallet',
        icon: 'https://wallet.example.org/icon.png',
      },
      provider: ethereum,
    },
  })
);
```